### PR TITLE
optimize computation and memory

### PR DESCRIPTION
**Pull request by felicepantaleo**
This PR should improve the overhead of looping over tiles by avoiding multiple copies

### CODE REVIEW

**Review by jmcarcell — COMMENTED** [review]
Looks good. Maybe @ebrondol wants to comment on this.
Did you measure if there is any improvement? I would guess that not much since the copies were in the outer parts of the for loops, unless there is a huge number of points (?)

**Comment by ebrondol** [comment]
> Looks good. Maybe @ebrondol wants to comment on this. Did you measure if there is any improvement? I would guess that not much since the copies were in the outer parts of the for loops, unless there is a huge number of points (?)

Hi @jmcarcell , I am testing the PR currently. The preliminary performance looks much better than before computing time wise, so I am confident it can be marged soon.

File: include/LayerTiles.h [hunk 1]
```diff
@@ -118,6 +118,10 @@ class LayerTiles_T {
     std::vector<int>& operator[](int globalBinId) {
       return layerTiles_[globalBinId];
     }
+    
+    const std::vector<int>& operator[](int globalBinId) const {
+      return layerTiles_[globalBinId];
+    }
```
[the remark this answers is not in the document]
Comment (felicepantaleo) [reply include/LayerTiles.h:124]: for `non-const` objects, you might want to be able to access a tile and modify it

**Comment by felicepantaleo** [comment]
> Looks good. Maybe @ebrondol wants to comment on this. Did you measure if there is any improvement? I would guess that not much since the copies were in the outer parts of the for loops, unless there is a huge number of points (?)

well, if every time you copy all the points in a layer, the overhead of creating tiles becomes bigger than the clustering itself and you enter the paradox in which coarser tiling gives better performance.

**Comment by ebrondol** [comment]
image
Indeed the timing is doing much better now. 
I will merge the PR.

File: src/CLUEAlgo.cc [hunk 2]
```diff
@@ -170,6 +170,7 @@ void CLUEAlgo_T<TILES>::findAndAssignClusters(){
 
   // find cluster seeds and outlier
   std::vector<int> localStack;
+  localStack.reserve(10);
```
[the remark this answers is not in the document]
Comment (ebrondol) [reply src/CLUEAlgo.cc:173]: pinging @felicepantaleo

**Review by ebrondol — APPROVED** [review]
The changes align with the expectation of about a factor of 50-100 improvements in computing time.

**Comment by ebrondol** [comment]
The tests I run locally worked out fine. 
I would rather wait until Issue #45 is closed and the CI is running again before merging it.

**Comment by felicepantaleo** [comment]
Thank you, Erica. Sure, I'm not in a hurry to have it merged.

**Comment by ebrondol** [comment]
Just for reference, I have produced the timing plots again with the latest commits from @felicepantaleo :
Screenshot from 2023-07-20 11-22-59
(disregard the label "before PR#43" because it is outside the x-axis region)

**Comment by ebrondol** [comment]
I have tested this PR extensively with other detectors and I think it can be merged also before the CI is fixed (still under progress and being monitored in #44 ).

**Pull request merged**